services/rest/websocket/connection/table.go

```diff
@@ -44,11 +44,26 @@ func (t *Table) Store(c identification.Identifier) error {
 	if _, ok := t.connMap[c]; ok {
 		return errConnDuplicated
 	}
-	t.connMap[c] = struct{}{}
+	t.connMap[c] = make(map[string]struct{})
 	t.counter[c.Group] = t.counter[c.Group] + 1
 	return nil
 }
 
+func (t *Table) StoreEvent(c identification.Identifier, eventId string) {
+	t.m.Lock()
```

Comment:
I presume, we wouldn't need to synchornize as the dedupe is on best effort basis and we can trade-off to save performance. Do you have string reason to do this?

Reply:
My reason: 
Because a single open connection can continue to send events, and because this list of event-Id's are mapped to a single key/user, updating this key for every incoming event may result in a `fatal` error, so I think we need the lock.

Reply:
I see you already are checking and updating only if absent. So this should be fine.
This will degrade performance if we fetch locks for every batch. Could we avoid this?

Reply:
I added the new benchmark test which is failing If I run with the -race detector and with the `no-locks`

```
❯ go test -benchmem -run=^$ -bench ^BenchmarkStoreBatch$ github.com/odpf/raccoon/services/rest/websocket/connection -tags dynamic -race
goos: darwin
goarch: arm64
pkg: github.com/odpf/raccoon/services/rest/websocket/connection
BenchmarkStoreBatch-10    	==================
WARNING: DATA RACE
Write at 0x00c000095200 by goroutine 21:
  runtime.mapaccessK()
      /usr/local/go/src/runtime/map.go:518 +0x1ec
  github.com/odpf/raccoon/services/rest/websocket/connection.(*Table).Store()
```

Reply:
to Improve the perf I have changed the `HasBatch` to have only read lock

Reply:
FYI: If I remove the lock, the code will throw the fatal error 
`fatal error: concurrent map read and map write`

![image](https://user-images.githubusercontent.com/5475978/199967680-bac34f4a-4070-4235-b438-6c3135ed08ec.png)

Reply:
We could proceed with what we have and have benchmarks captured on this new code,

Resolving this now.